morango/sync/syncsession.py

```diff
@@ -376,42 +384,139 @@ def _pull_record_chunk(self, chunk_size, transfer_session):
         return self.session.get(self.urlresolve(api_urls.BUFFER), params=params)
 
 
+class SyncSignal(object):
+    """
+    Helper class for firing signals from the sync client
+    """
+
+    def __init__(self, **kwargs_defaults):
+        """
+        Default keys/values that the signal consumer can depend on being present.
+        """
+        self._handlers = []
+        self._defaults = kwargs_defaults
+
+    def connect(self, handler):
+        """
+        Adds a callable handler that will be called when the signal is fired.
+
+        :type handler: function
+        """
+        self._handlers.append(handler)
+
+    def fire(self, **kwargs):
+        """
+        Fires the handler functions connected via `connect`.
+        """
+        fire_kwargs = self._defaults.copy()
+        fire_kwargs.update(kwargs)
+
+        for handler in self._handlers:
+            handler(**fire_kwargs)
+
+
+class SyncSignalGroup(SyncSignal):
+    """
+    Breaks down a signal into `started`, `in_progress`, and `completed` stages. The
+    `kwargs_defaults` are passed through to each signal stage.
+    """
+
+    started = SyncSignal()
+    """The started signal, which will be fired at the beginning of the procedure."""
+    in_progress = SyncSignal()
+    """The in progress signal, which should be fired at least once during the procedure."""
+    completed = SyncSignal()
+    """The completed signal, which should be fired at the end of the procedure"""
+
+    def __init__(self, **kwargs_defaults):
+        super(SyncSignalGroup, self).__init__(**kwargs_defaults)
+        self.started = SyncSignal(**kwargs_defaults)
+        self.in_progress = SyncSignal(**kwargs_defaults)
+        self.completed = SyncSignal(**kwargs_defaults)
+
+        self.started.connect(self.fire)
+        self.in_progress.connect(self.fire)
+        self.completed.connect(self.fire)
+
+    def send(self, **kwargs):
+        """
+        Context manager helper that will signal started and fired when entered and exited,
+        and it'll fire those with the `kwargs`.
+
+        :rtype: SyncSignalGroup
+        """
+        defaults = self._defaults.copy()
+        defaults.update(kwargs)
+        context_group = SyncSignalGroup(**defaults)
+        context_group.started.connect(self.started.fire)
+        context_group.in_progress.connect(self.in_progress.fire)
+        context_group.completed.connect(self.completed.fire)
+        return context_group
+
+    def __enter__(self):
+        """
+        Fires the `started` signal.
+        """
+        self.started.fire()
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """
+        Fires the `completed` signal.
+        """
+        self.completed.fire()
```

Comment:
At the moment, I have no requirements on how to handle or report an error case. We could avoid firing `complete` this if there's an error?

Reply:
I think still firing `completed` would be best, and then we should have the Kolibri code check what the status is when it's updating what it shows. I guess it may also be handled in parallel by the job erroring out.